core-rust/state-manager/src/protocol/protocol_state.rs

```diff
@@ -18,76 +19,173 @@ impl NodeProtocolUpdateExecutor {
-            .create_batch_generator_raw(&self.network, self.database.clone(), overrides);
-
-        let transactions_and_scenarios = WithScenariosNodeBatchGenerator {
-            base_batch_generator: protocol_update_transactions.deref(),
-            scenario_names: self
-                .scenarios_execution_config
-                .to_run_after_protocol_update(protocol_version),
-        };
-
-        for batch_index in from_batch_idx..transactions_and_scenarios.batch_count() {
-            self.system_executor.execute_protocol_update_action(
-                protocol_version,
-                batch_index,
-                transactions_and_scenarios.generate_batch(batch_index),
-            );
+        let resolved = protocol_version.validate().unwrap_or_else(|err| {
+            panic!("{protocol_version:?} is not a supported protocol version: {err:?}")
+        });
+        let protocol_update_definition = resolved.definition();
+        let update_generator = protocol_update_definition.create_update_generator_raw(
+            ProtocolUpdateContext {
+                network: &self.network,
+                database: &self.database,
+                genesis_data_resolver: &self.genesis_data_resolver,
+                scenario_config: &self.scenarios_execution_config,
+            },
+            overrides,
+        );
+        let config_hash = update_generator.config_hash();
+
+        let batch_groups = update_generator.batch_groups();
+        let total_batch_groups = batch_groups.len();
+        let remaining_batch_groups = batch_groups
+            .into_iter()
+            .enumerate()
+            .skip(from_batch_group_index);
+
+        for (batch_group_index, batch_group) in remaining_batch_groups {
+            let batch_group_number = batch_group_index + 1;
+            let batch_group_name = batch_group.batch_group_name();
+            let is_final_batch_group = batch_group_index == total_batch_groups - 1;
+            let start_at_batch = if batch_group_index == from_batch_group_index
+                && from_batch_index > 0
+            {
+                info!("[UPDATE: {protocol_version}] Continuing {batch_group_name} (batch group {batch_group_number}/{total_batch_groups})");
+                from_batch_index
+            } else {
+                info!("[UPDATE: {protocol_version}] Commencing {batch_group_name} (batch group {batch_group_number}/{total_batch_groups})");
+                0
+            };
+            let batches = {
+                // In a separate block to ensure the database lock guard is dropped promptly.
+                batch_group.generate_batches(self.database.lock().deref())
+            };
+            let total_batches = batches.len();
+            let remaining_batches = batches.into_iter().enumerate().skip(start_at_batch);
+            for (batch_index, batch_generator) in remaining_batches {
+                let batch_number = batch_index + 1;
```

Comment:
Doesn't seem right, shouldn't the `batch_number` be `start_at_batch + batch_index`?

Reply:
looks right to me :thinking: the list is enumerated before `skip`